tsdb/engine/wal/wal.go

```diff
@@ -0,0 +1,1013 @@
+func (w *Log) pointsToPartitions(points []tsdb.Point) map[*partition][]tsdb.Point {
+	m := make(map[*partition][]tsdb.Point)
+	for _, p := range points {
+		pp := w.walPartition(p.Key())
+		m[pp] = append(m[pp], p)
+	}
+	return m
+}
+
+// openPartitionFiles will open all partitions and read their segment files
+func (w *Log) openPartitionFiles() error {
+	results := make(chan error, len(w.partitions))
+	for _, p := range w.partitions {
+
+		go func(p *partition) {
+			fileNames, err := p.segmentFileNames()
+			if err != nil {
+				results <- err
+				return
+			}
+			for _, n := range fileNames {
+				entries, err := p.readFile(n)
+				if err != nil {
+					results <- err
+					return
+				}
+				for _, e := range entries {
+					p.addToCache(e.key, e.data, e.timestamp)
+				}
+			}
+			results <- nil
+		}(p)
+	}
+
+	for i := 0; i < len(w.partitions); i++ {
+		err := <-results
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// Close will finish any flush that is currently in process and close file handles
+func (w *Log) Close() error {
+	// stop the autoflushing process so it doesn't try to kick another one off
+	if w.closing != nil {
+		close(w.closing)
+		w.closing = nil
+	}
+
+	w.wg.Wait()
+
+	w.mu.Lock()
+	defer w.mu.Unlock()
+
+	// clear the cache
+	w.partitions = nil
+
+	return w.close()
+}
+
+// close all the open Log partitions and file handles
+func (w *Log) close() error {
+	for _, p := range w.partitions {
+		if err := p.Close(); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// triggerAutoFlush will flush and compact any partitions that have hit the thresholds for compaction
+func (l *Log) triggerAutoFlush() {
+	l.mu.RLock()
+	defer l.mu.RUnlock()
+	for _, p := range l.partitions {
+		if p.shouldFlush(l.MaxSeriesSize, l.ReadySeriesSize, l.CompactionThreshold, l.PartitionSizeThreshold) {
+			if err := p.flushAndCompact(l.IndexWriter, l.MaxSeriesSize, l.ReadySeriesSize); err != nil {
+				l.logger.Printf("error flushing partition %d: %s\n", p.id, err)
+			}
+		}
+	}
+}
+
+// autoflusher waits for notification of a flush and kicks it off in the background.
+// This method runs in a separate goroutine.
+func (w *Log) autoflusher(closing chan struct{}) {
+	defer w.wg.Done()
+
+	for {
+		// Wait for close or flush signal.
+		select {
+		case <-closing:
+			return
+		case <-w.flushCheckTimer.C:
+			w.triggerAutoFlush()
+		case <-w.flush:
+			if err := w.Flush(); err != nil {
+				w.logger.Printf("flush error: %s", err)
+			}
+		}
+	}
+}
+
+// walPartition returns the partition number that key belongs to.
+func (l *Log) walPartition(key []byte) *partition {
+	h := fnv.New64a()
+	h.Write(key)
+	id := uint8(h.Sum64()%l.PartitionCount + 1)
+	p := l.partitions[id]
+	if p == nil {
+		if p, err := newPartition(id, l.path, l.SegmentSize); err != nil {
+			panic(err)
+
+		} else {
+			l.partitions[id] = p
+		}
+	}
+	return p
+}
```

Comment:
Can you move this function and the previous couple functions up to where the `Log` functions are? Right now it's after a bunch of `partition` and `segmentFile` functions.